From 98ad3653ca853932e2c754315540ba18a09c495c Mon Sep 17 00:00:00 2001
From: Viktor Ashirov <vashirov@redhat.com>
Date: Sat, 8 Nov 2025 01:05:15 +0100
Subject: [PATCH] topology: fix memory leaks

`agmt_attr_val`, `targetHost` and internal search results pblock were
not freed.

Related: https://pagure.io/freeipa/issue/9895
Signed-off-by: Viktor Ashirov <vashirov@redhat.com>
Reviewed-By: Rob Crittenden <rcritten@redhat.com>
Reviewed-By: Alexander Bokovoy <abbra@users.noreply.github.com>
---
 daemons/ipa-slapi-plugins/topology/topology_util.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/daemons/ipa-slapi-plugins/topology/topology_util.c b/daemons/ipa-slapi-plugins/topology/topology_util.c
index 3f9bf137fdcd398e12dfa7bfbcd9ccde3d773c99..cd3e70910d54bb4f1853ef68ea98d69f2f97ba7e 100644
--- a/daemons/ipa-slapi-plugins/topology/topology_util.c
+++ b/daemons/ipa-slapi-plugins/topology/topology_util.c
@@ -678,6 +678,7 @@ ipa_topo_util_update_agmt_list(TopoReplica *conf, TopoReplicaSegmentList *repl_s
                                               mattrs[i],
                                               segm_attr_val);
                     }
+                    slapi_ch_free_string(&agmt_attr_val);
 
                 }
             }
@@ -711,10 +712,10 @@ ipa_topo_util_update_agmt_list(TopoReplica *conf, TopoReplicaSegmentList *repl_s
                 ipa_topo_cfg_segment_set_visited(conf, topo_segm);
             }
         }
+        slapi_ch_free_string(&targetHost);
 
         repl_agmt = entries[++nentries];
     }
-    slapi_free_search_results_internal(pb);
 
 update_only:
     /* check if segments not covered by agreement exist
@@ -724,6 +725,7 @@ update_only:
                                     ipa_topo_get_plugin_hostname());
 
 error_return:
+    slapi_free_search_results_internal(pb);
     slapi_ch_free_string(&filter);
     slapi_pblock_destroy(pb);
     return rc;
-- 
2.52.0

